### PR TITLE
[apple-reminders] Display due time in menu-bar reminder titles

### PR DESCRIPTION
**Pull request by isabhiii**
## Checklist

- [x] I read the extension guidelines
- [x] I read the documentation about publishing
- [x] I ran `npm run build`
- [x] I checked that files in the `assets` folder are used
- [x] I checked that assets used in the `README` are located outside the metadata folder if they were not generated with our [metadata tool](https://developers.raycast.com/basics/prepare-an-extension-for-store#how-to-use-it)

### CODE REVIEW

**Review by 0xdhrv — COMMENTED** [review]
Hi @isabhiii 👋

Thanks for the contribution 🔥

Could you look into the suggestions from Greptile?

**Comment by isabhiii** [comment]
Hi @0xdhrv 👋🏻

Applied the suggestion! Let me know if there's anything else.

**Comment by isabhiii** [comment]
@0xdhrv How long until it will be reviewed? Are any more changes required?

File: extensions/apple-reminders/package.json [hunk 1]
```diff
@@ -159,15 +160,6 @@
           "default": false,
           "required": false
         },
-        {
```
Comment (0xdhrv) [inline extensions/apple-reminders/package.json:162]: **issue (blocking):** Removing this preference forces due-date sorting for everyone.

That changes behavior beyond deleting an obsolete setting, especially because the removed preference previously allowed users to control this sorting. Could we preserve the existing choice or provide an alternative before removing it?

File: extensions/apple-reminders/src/helpers.ts [hunk 2]
```diff
@@ -8,6 +8,22 @@ export function isFullDay(date: string) {
   return /^\d{4}-\d{2}-\d{2}$/.test(date);
 }
 
+export function formatReminderTime(reminder: Reminder): string {
```
Comment (0xdhrv) [inline extensions/apple-reminders/src/helpers.ts:11]: **issue** (blocking): Handle all-day reminders before formatting the time

Could we use `isFullDay(reminder.dueDate)` in `formatReminderTime` and return an appropriate all-day value before calling `Intl.DateTimeFormat`? Without that check, date-only reminders may be parsed with an unintended time zone and displayed as a misleading time.

**Review by 0xdhrv — CHANGES_REQUESTED** [review]
Hey @isabhiii 👋 

I have added a few comments for you to address.

I'm looking forward to testing this extension again 🔥 

Feel free to contact me here or at [Slack](https://www.raycast.com/community) if you have any questions.

---

I converted this PR into a draft until it's ready for the review, please press the button `Ready for review` when it's ready and we'll have a look 😊

File: extensions/apple-reminders/src/helpers.ts (hunk 2 above)
Comment (isabhiii) [reply extensions/apple-reminders/src/helpers.ts:11]: Resolved!

**Comment by isabhiii** [comment]
@0xdhrv It is ready for the review 😊

**Comment by isabhiii** [comment]
@0xdhrv brody, it is done and have been submitted. please have a look ;)

File: extensions/apple-reminders/package.json [hunk 3]
```diff
@@ -164,7 +173,7 @@
           "type": "checkbox",
           "label": "Sort menu bar reminders by time",
           "description": "When enabled, the reminders in the menu bar will be sorted by time.",
-          "default": false,
+          "default": true,
```
Comment (0xdhrv) [inline extensions/apple-reminders/package.json]: **issue (blocking):** Could we keep the existing default for `sortMenuBarRemindersByDueDate` as `false`?

Changing it to `true` means users who haven't explicitly configured this preference will have their menu-bar reminders reordered after updating the extension. Since this PR is focused on displaying due times, I don't think we should also change the existing default sorting behavior.

Keeping the current default would make this update backward-compatible while still allowing users to opt into sorting by time.

```suggestion
          "default": false,
```

File: extensions/apple-reminders/src/menu-bar.tsx [hunk 4]
```diff
@@ -216,18 +225,19 @@ export default function Command() {
...
-                      : reminder.title,
-                    reminder.priority,
-                  ),
-                )}
+                title={`${timePrefix}${addPriorityToTitle(
+                  displayListTitleForMenuBarReminders
+                    ? addListTitle(reminder.title, reminder.list?.title)
+                    : reminder.title,
+                  reminder.priority,
+                )}`}
```
Comment (0xdhrv) [inline extensions/apple-reminders/src/menu-bar.tsx]: **issue:** Could we keep the existing `truncate()` behavior here?

This change removes the truncation that was previously applied to menu-bar reminder titles. With the new time prefix (and potentially the list title as well), long reminder names can now make the menu unnecessarily wide.

The due time can be added while preserving the existing truncation behavior.

```suggestion
                title={truncate(
                  `${timePrefix}${addPriorityToTitle(
                    displayListTitleForMenuBarReminders
                      ? addListTitle(reminder.title, reminder.list?.title)
                      : reminder.title,
                    reminder.priority,
                  )}`,
                )}
```

**Review by 0xdhrv — CHANGES_REQUESTED** [review]
There seem to be quite a few unrelated changes in `menu-bar.tsx` alongside the due-time feature—for example changes to the delete flow, empty state, icons, attached-URL toast, priority/date menu ordering, list selector, and the Open Reminders action.

Could we keep this PR focused on displaying the due time and revert the unrelated behavioral/UI changes? That would make the change much easier to review and reduce the chance of introducing regressions in existing menu-bar functionality.

---

I converted this PR into a draft until it's ready for the review, please press the button `Ready for review` when it's ready and we'll have a look 😊

Request a new review when you are ready. Feel free to contact me here or at [Slack](https://www.raycast.com/community) if you have any questions.

**Comment by isabhiii** [comment]
@0xdhrv  Hey Dhruv, suggested changes are done, have a look :)

File: extensions/apple-reminders/src/menu-bar.tsx [hunk 5]
```diff
@@ -269,119 +279,113 @@ export default function Command() {
...
+                    onAction={() => setDueDate(reminder.id, endOfWeek(startOfToday(), { weekStartsOn: 1 }))}
                     icon={Icon.Calendar}
-                    onAction={() => setDueDate(reminder.id, startOfWeek(addWeeks(now, 1), { weekStartsOn: 1 }))}
                   />
                   <MenuBarExtra.Item
                     title="No Due Date"
-                    icon={Icon.XMarkCircle}
                     onAction={() => setDueDate(reminder.id, null)}
+                    icon={Icon.XMarkCircle}
                   />
```
Comment (0xdhrv) [inline extensions/apple-reminders/src/menu-bar.tsx:323]: **issue (blocking):** Please revert this due-date submenu. None of it is needed to display due time.

In particular:
- `This Weekend` / `Next Week` are swapped
- icons/labels changed (`This Week-End` → `This Weekend`, `Icon.Clock`/`Icon.Sunrise`/`Icon.ArrowClockwise` → calendar/sun)
- `endOfWeek(now, { weekStartsOn: 1 })` became `endOfWeek(startOfToday(), { weekStartsOn: 1 })`

That last one may be equivalent for most dates, but it is unrelated churn. Please restore the existing action order, copy, icons, and `now`-based date math.

File: extensions/apple-reminders/src/menu-bar.tsx [hunk 6]
```diff
@@ -269,119 +279,113 @@ export default function Command() {
...
+                      icon: { source: Icon.Trash, tintColor: Color.Red },
+                      primaryAction: {
+                        title: "Delete",
+                        style: Alert.ActionStyle.Destructive,
+                        onAction: () => deleteReminder(reminder),
+                      },
+                    })
                   }
+                  icon={Icon.Trash}
                 />
```
Comment (0xdhrv) [inline extensions/apple-reminders/src/menu-bar.tsx:366]: **issue (blocking):** Please restore the existing delete confirmation.

This replaces an awaited `confirmAlert` plus an Option-key `alternate` immediate-delete action with a `primaryAction` callback. That removes an intentional power-user shortcut and changes an established destructive interaction without a reason tied to due-time display.

Also please keep the ellipsis character (`Delete Reminder…`) rather than `...`.

File: extensions/apple-reminders/src/menu-bar.tsx [hunk 7]
```diff
@@ -391,12 +395,11 @@ export default function Command() {
           }
         />
       </MenuBarExtra.Section>
-
       <MenuBarExtra.Section>
         <MenuBarExtra.Item
           title="Open Reminders"
           icon={{ fileIcon: REMINDERS_FILE_ICON }}
-          onAction={() => open("home", "com.apple.reminders")}
+          onAction={() => open("x-apple-reminderkit://", "com.apple.reminders")}
```
Comment (0xdhrv) [inline extensions/apple-reminders/src/menu-bar.tsx]: **issue (blocking):** Please keep `open("home", "com.apple.reminders")`.

Switching to `x-apple-reminderkit://` is unrelated to due-time display, and custom URL schemes can be brittle across macOS versions. The existing `home` target is the known-good way to launch Reminders.

File: extensions/apple-reminders/src/menu-bar.tsx [hunk 8]
```diff
@@ -269,119 +279,113 @@ export default function Command() {
...
-              key={list.id}
-              title={list.title}
-              onAction={() => handleListChange(list.id)}
-              icon={{ source: Icon.Circle, tintColor: list.color }}
+              key={l.id}
+              title={l.title}
+              icon={{ source: Icon.Circle, tintColor: l.color }}
+              onAction={() => handleListChange(l.id)}
             />
           ))}
```
Comment (0xdhrv) [inline extensions/apple-reminders/src/menu-bar.tsx:395]: **issue (blocking):** Please restore the empty state and the existing list selector.

This hunk also drops `You don't have any reminders.`, changes the All-lists icon from `Icon.Tray` to `Icon.Circle`, and calls `handleListChange()` instead of `handleListChange(undefined)`. None of that belongs in a due-time display PR.

File: extensions/apple-reminders/src/menu-bar.tsx [hunk 9]
```diff
@@ -269,119 +279,113 @@ export default function Command() {
...
-                    icon={getPriorityIcon("low")}
                     onAction={() => setPriority(reminder.id, "low")}
+                    icon={getPriorityIcon("low")}
+                  />
+                  <MenuBarExtra.Item
+                    title="None"
+                    onAction={() => setPriority(reminder.id, null)}
+                    icon={Icon.XMarkCircle}
                   />
                 </MenuBarExtra.Submenu>
```
Comment (0xdhrv) [inline extensions/apple-reminders/src/menu-bar.tsx:343]: **issue:** Please restore the existing priority submenu (None first, no extra `XMarkCircle` on None).

Reordering this menu is unrelated to displaying due time.

File: extensions/apple-reminders/src/components/ReminderListItem.tsx [hunk 10]
```diff
@@ -48,10 +48,27 @@ export default function ReminderListItem({
...
+    } else if (displayDueDateWithTime) {
+      // Show exact time (e.g. 10:00 AM)
+      const dateObj = new Date(dueDate);
+      valueText = new Intl.DateTimeFormat(undefined, {
+        hour: "2-digit",
+        minute: "2-digit",
+      }).format(dateObj);
+    } else {
+      valueText = formatDistanceToNow(dueDate, { addSuffix: true });
+    }
```
Comment (0xdhrv) [inline extensions/apple-reminders/src/components/ReminderListItem.tsx:68]: **issue:** Exact-time formatting is duplicated here instead of reusing `formatReminderTime` from `helpers.ts`.

This preference also only affects list accessories — the menu bar always prefixes exact time, ignoring `displayDueDateWithTime`. Please either:

1. Keep the preference in this PR, call `formatReminderTime(reminder)` here, and document the menu-bar vs list split in the changelog, or
2. Drop the preference and these `ReminderListItem` edits so this PR is menu-bar-only.

File: extensions/apple-reminders/package.json [hunk 11]
```diff
@@ -49,6 +50,14 @@
...
           "description": "When enabled, today's reminders will be grouped by the time of day they are due.",
           "default": false
+        },
+        {
+          "name": "displayDueDateWithTime",
+          "type": "checkbox",
+          "required": false,
+          "label": "Display Exact Due Time",
+          "description": "When enabled, list items show the exact time (e.g. 10:00 AM) instead of relative time (e.g. 'in 2 hours').",
+          "default": false
```
Comment (0xdhrv) [inline extensions/apple-reminders/package.json:60]: **issue:** This preference is on **My Reminders**, but the PR title/changelog describe a menu-bar feature. Menu-bar titles always show exact time and never read `displayDueDateWithTime`.

Please either keep this only if list exact-time is intentionally in scope (and update the changelog accordingly), or remove it from this PR.

File: extensions/apple-reminders/src/helpers.ts [hunk 12]
```diff
@@ -8,6 +8,22 @@ export function isFullDay(date: string) {
...
+  const date = new Date(reminder.dueDate);
+  if (isNaN(date.getTime())) {
+    return "";
+  }
+
+  return new Intl.DateTimeFormat(undefined, {
+    hour: "2-digit",
+    minute: "2-digit",
+  }).format(date);
+}
```
Comment (0xdhrv) [inline extensions/apple-reminders/src/helpers.ts]: This helper is the right place for the feature — please keep it, reuse it from `ReminderListItem` if the list preference stays, and add small tests for:

- all-day dates (`YYYY-MM-DD`) → `""`
- valid timed dates → locale time string
- malformed dates → `""`

**Review by 0xdhrv — CHANGES_REQUESTED** [review]
Hey @isabhiii 👋

Thanks for iterating on this. The due-time prefix itself looks good, but **please do not merge this as-is** — the PR is still mixing a focused display change with unrelated menu-bar behavior changes, which makes review and rollback hard.

### Please revert unrelated `menu-bar.tsx` edits

This is nominally a due-time display feature, but `menu-bar.tsx` also changes:

- Delete-reminder confirmation flow (and removes the Option-key immediate-delete alternate)
- Due-date quick actions, labels, icons, and menu order (`This Weekend` / `Next Week`)
- Priority menu order and icons
- Attached-URL failure toast content
- Empty-state removal (`You don't have any reminders.`)
- List-selector icons and `handleListChange()` invocation
- “Open Reminders” deep link (`home` → `x-apple-reminderkit://`)

Please revert all of those. Keep only:

- `formatReminderTime`
- menu-bar title prefixing for the top-level title and reminder submenu titles
- changelog / contributor metadata as appropriate

### Decide whether the list preference belongs here

**Display Exact Due Time** currently affects `ReminderListItem` only. The menu bar always prefixes exact time via `formatReminderTime`, regardless of that preference.

That can be acceptable if the intent is “always show exact time in the menu bar,” but then naming/changelog should distinguish:

- Menu-bar due-time prefix — always enabled
- List accessory exact vs. relative time — configurable

If the preference stays, reuse `formatReminderTime` in `ReminderListItem` instead of duplicating `Intl.DateTimeFormat`. If it does not belong in this PR, remove the preference and the `ReminderListItem` edits entirely.

### Tests / verification

Please add small tests for `formatReminderTime` (all-day dates, valid timed dates, malformed dates) and manually verify long titles, all-day reminders, completed reminders, and a non-English system locale.

---

I converted this PR into a draft until it's ready for review. Please press **Ready for review** when the unrelated `menu-bar.tsx` changes are reverted and we can take another look 😊

Feel free to contact me here or at [Slack](https://www.raycast.com/community) if you have any questions.

File: extensions/apple-reminders/src/menu-bar.tsx (hunk 6 above)
Comment (isabhiii) [reply extensions/apple-reminders/src/menu-bar.tsx:366]: Reverted to match upstream and kept only the due-time prefix and title truncation.

File: extensions/apple-reminders/src/components/ReminderListItem.tsx (hunk 10 above)
Comment (isabhiii) [reply extensions/apple-reminders/src/components/ReminderListItem.tsx:68]: Updated to reuse formatReminderTime(reminder)

File: extensions/apple-reminders/src/helpers.ts (hunk 12 above)
Comment (isabhiii) [reply extensions/apple-reminders/src/helpers.ts]: Updated and added unit tests covering all edge cases in tests/helpers.test.ts

**Comment by isabhiii** [comment]
Hi @0xdhrv 👋

Thanks for the detailed review! I've updated the PR with all the requested changes:

1. **Reverted unrelated `menu-bar.tsx` changes**: Restored the original delete confirmation and Option-key alternate delete, due-date submenu ordering/icons/date math, priority submenu order, empty state, list selector, and the `open("home", "com.apple.reminders")` action while keeping the due-time prefix and preserving title truncation.
2. **Reused `formatReminderTime`**: Updated `ReminderListItem.tsx` to directly call `formatReminderTime(reminder)` when `displayDueDateWithTime` is enabled.
3. **Preferences & Defaults**: Restored `sortMenuBarRemindersByDueDate` default to `false` and kept `displayDueDateWithTime` (`default: false`).
4. **Unit Tests**: Added test coverage in `tests/helpers.test.ts` for `formatReminderTime` covering all-day dates, valid timed dates, malformed dates, and missing/null dates.
5. **Changelog**: Updated `CHANGELOG.md` to document both the menu bar due-time display and the list exact-time preference.

Everything has been validated and linted (`ray lint`). Ready for another review! 😊

**Review by isabhiii — COMMENTED** [review]
Hi @0xdhrv 👋

I've pushed the latest updates to address the remaining points:

- **Unicode-safe truncation**: Updated `truncate()` in `helpers.ts` using `Array.from()` to ensure emojis and multi-byte characters are preserved without splitting when truncating menu bar titles.
- **Shared helper tests & Raycast mocks**: Added `@raycast/api` module mocks (`tests/register-mocks.mjs` and `tests/mocks/raycast-api.mjs`) so unit tests in `tests/helpers.test.ts` can import `formatReminderTime` and `truncate` directly from `src/helpers.ts`.
- **Test coverage**: Added test cases for all-day dates, valid timed dates, malformed dates, and Unicode/emoji string truncation (all 7 tests passing).
- **Test script & cleanup**: Added `npm test` script to `package.json`, switched to type-only imports in helpers, and added the changelog entry.

All tests and lint checks (`ray lint`) pass cleanly. Ready for your review! 😊

**Review by 0xdhrv — APPROVED** [review]
Looks good to me, approved ✅

**Comment by isabhiii** [comment]
> 🎉 🎉 🎉
> 
> We've rewarded your Raycast account with some credits. You will soon be able to exchange them for some swag.

In brightest day, in blackest night,
**No merge conflict** shall escape my sight.
Let those who fear **Unicode edge-cases bite**,
Beware my power, Raycast swag light! 🤜🏻 ✳️

Thanks @0xdhrv for guiding the ring through the review gauntlet! 😄
